### PR TITLE
combined emd_token and attn training in single stage, added flash_attn_2 argument in training script

### PR DESCRIPTION
**Pull request by ajaymin28**
### 🚀 Summary

Fixes: https://github.com/ariG23498/gemma3-object-detection/issues/37
Addresses: https://github.com/ariG23498/gemma3-object-detection/issues/39

This PR combines the training of **emb_token** and **attn layers** into a single stage, rather than training them separately in two stages.

The original approach involved first training the **embed_tokens** layer, freezing it, and then training the **attn layers** in a second stage. However, this two-stage method doesn't yield satisfactory results, see the results below.

Previously:

```python
run_training_phase(model, processor, cfg, train_dataloader, train_keys=["embed_tokens"], phase_name="embed_only")
# embed_tokens learned

run_training_phase(model, processor, cfg, train_dataloader, train_keys=["embed_tokens", "attn"], phase_name="embed_attn")
# embed_tokens updated/disrupted
``` 

Now: 

```python
# embed_tokens and attn in single stage
run_training_phase(model, processor, cfg, train_dataloader, train_keys=["embed_tokens", "attn"], phase_name="embed_attn_embed_tokens")
```

---

### 896 <locxxxx> tokens instead of 1024 :question: 
My assumption was that matching the location tokens with image size would yield better results, but that's not true. See below in **Observations**. I have opened a separate issue here: https://github.com/ariG23498/gemma3-object-detection/issues/39, which I will close with appropriate comments. 

---

##  Why not embed_tokens training followed by attn training in two stages :question: 

As experimented, please find below the results for two-stage and one-stage training. 

### 📊 Results

<table>
  <tr>
    <th>Two stage(Previously) 1024 loc tokens</th>
    <th>One stage(Now) 1024 loc tokens</th>
    <th>One stage <br> 896 loc tokens instead of 1024 <br><span style='font-size:12px'>(addresses issue: <a href="https://github.com/ariG23498/gemma3-object-detection/issues/39">#39</a>)</span></th>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

### 🤔 Observations

- Single-stage training consistently yields better results, as demonstrated above. This is a positive outcome since training in a single stage reduces the overall training effort compared to the two stage training.

- Longer training durations are beneficial: Models trained for 10 epochs outperform those trained for only 2 epochs, where performance remains poor.

- Using 896 bins (location tokens) i.e., matching the input size, did not lead to improved results, contrary to my initial assumption (see [this issue](https://github.com/ariG23498/gemma3-object-detection/issues/39)). This finding suggests that matching the the number of tokens may not enhance performance, although it could be worth exploring even higher token counts in future experiments.

### :bug: Issues

For saving the models, I faced an issue with safe serialization of tensors, so the current models are saved using the following code. check this issue: https://github.com/huggingface/transformers/issues/27613#issuecomment-1848645557

```python
model.push_to_hub(cfg.checkpoint_id, safe_serialization=False)
processor.push_to_hub(cfg.checkpoint_id, safe_serialization=False)
```

### Loss graph

image

- I think we should employ a learning rate scheduler here: Warmup for first few steps and then slowly cool down. 

### :smile: Checkpoints

- [ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-1024-single-stage](https://huggingface.co/ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-1024-single-stage)
- [ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-1024-two-stage](https://huggingface.co/ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-1024-two-stage)
- [ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-896-single-stage](https://huggingface.co/ajaymin28/gemma-3-4b-pt-obj-det-loc-tok-896-single-stage)

### :dart: Future work

I think it's a promising result so far, but I believe there are a few areas that could further improve performance:

- We should consider which additional components to fine-tune beyond just **emb_tokens** and **attn**. I have opened an issue to discuss this in detail: https://github.com/ariG23498/gemma3-object-detection/issues/41

- ~~We may need to switch to a chat template, or at least verify that user and assistant messages are in the correct order. Please refer to this issue for more information: https://github.com/ariG23498/gemma3-object-detection/issues/40~~ Chat templates are applicable when we use instruction-tuned models(it), not pretrained(pt).

Let me know if you have any comments 😄

### CODE REVIEW

**Review by sergiopaniego — APPROVED** [review]
Thanks for the update and specially for the detailed PR!! 👏 
LGTM!

**Comment by sergiopaniego** [comment]
Would you be up for updating the model in the HF Space with the new one? 🚀 
https://huggingface.co/spaces/ariG23498/gemma3-license-plate-detection
